packages/sync/src/class-replicastore.php

```diff
@@ -1266,171 +1222,66 @@ public function checksum_histogram( $object_type, $buckets, $start_id = null, $e
-				$where_sql    = Settings::get_whitelisted_comment_meta_sql();
-				$object_count = $this->meta_count( $object_table, $where_sql, $start_id, $end_id );
-				$id_field     = 'meta_id';
-				break;
 			case 'terms':
-				$object_table = $wpdb->terms;
-				$object_count = $this->term_count();
-				$id_field     = 'term_id';
-				$where_sql    = '1=1';
+			case 'term_relationships':
+				$table = $object_type;
 				break;
-			case 'term_taxonomy':
-				$object_table = $wpdb->term_taxonomy;
-				$object_count = $this->term_taxonomy_count();
-				$id_field     = 'term_taxonomy_id';
-				$where_sql    = '1=1';
+			case 'post_meta':
+				$table = 'postmeta';
 				break;
-			case 'term_relationships':
-				$object_table = $wpdb->term_relationships;
-				$object_count = $this->term_relationship_count();
-				$id_field     = 'object_id';
-				$where_sql    = '1=1';
+			case 'comment_meta':
+				$table = 'commentmeta';
 				break;
 			default:
 				return false;
 		}
 
+		$checksum_table = new Table_Checksum( $table, $salt );
+		$range_edges    = $checksum_table->get_range_edges();
+
+		$object_count = $range_edges['item_count'];
+
 		$bucket_size     = (int) ceil( $object_count / $buckets );
 		$previous_max_id = 0;
 		$histogram       = array();
 
-		// This is used for the min / max query, while $where_sql is used for the checksum query.
-		$where = $where_sql;
-
-		if ( $start_id ) {
-			$where .= " AND $id_field >= " . (int) $start_id;
-		}
-
-		if ( $end_id ) {
-			$where .= " AND $id_field <= " . (int) $end_id;
-		}
-
 		do {
-			$result = $this->get_min_max_object_id(
-				$id_field,
-				$object_table,
-				$where . " AND $id_field > $previous_max_id",
-				$bucket_size
-			);
+			$ids_range = $checksum_table->get_range_edges( $previous_max_id, null, $bucket_size );
 
-			if ( null === $result->min || null === $result->max ) {
+			if ( empty( $ids_range['min_range'] ) || empty( $ids_range['max_range'] ) ) {
 				// Nothing to checksum here...
 				break;
 			}
 
 			// Get the checksum value.
-			$value = $this->table_checksum( $object_table, $columns, $id_field, $where_sql, $result->min, $result->max, $strip_non_ascii, $salt );
+			$batch_checksum = $checksum_table->calculate_checksum( $ids_range['min_range'], $ids_range['max_range'] );
 
-			if ( is_wp_error( $value ) ) {
-				return $value;
+			if ( is_wp_error( $batch_checksum ) ) {
+				return $batch_checksum;
 			}
 
-			if ( null === $result->min || null === $result->max ) {
-				break;
-			} elseif ( $result->min === $result->max ) {
-				$histogram[ $result->min ] = $value;
+			if ( $ids_range['min_range'] === $ids_range['max_range'] ) {
+				$histogram[ $ids_range['min_range'] ] = $batch_checksum;
 			} else {
-				$histogram[ "{$result->min}-{$result->max}" ] = $value;
+				$histogram[ "{$ids_range[ 'min_range' ]}-{$ids_range[ 'max_range' ]}" ] = $batch_checksum;
 			}
 
-			$previous_max_id = $result->max;
+			$previous_max_id = $ids_range['max_range'];
 		} while ( true );
```

Comment:
do we need to worry about memory usage here? should there be any clean up (stop_the_insanity) after x iterations?

Reply:
That's a worry on WPCOM more than on Jetpack sites. We need to test it with larger datasets and decide if we see any possible issues. We might need to add an action call here to be able to clean up memory while doing the loop.